File: src/adapters/persistence/WavEncoder.ts
/**
 * Utility functions to encode raw PCM 16-bit Mono audio bytes into standard RIFF WAV format.
 */
export function encodePcm16ToWav(pcmBytes: Uint8Array, sampleRate: number): Buffer {
  const numChannels = 1;
  const bitsPerSample = 16;
  const byteRate = (sampleRate * numChannels * bitsPerSample) / 8;
  const blockAlign = (numChannels * bitsPerSample) / 8;
  const dataSize = pcmBytes.byteLength;
  const headerSize = 44;
  const totalSize = headerSize + dataSize;

  const buffer = Buffer.alloc(totalSize);

  // RIFF Chunk Descriptor
  buffer.write("RIFF", 0); // ChunkID
  buffer.writeUInt32LE(36 + dataSize, 4); // ChunkSize
  buffer.write("WAVE", 8); // Format

  // fmt sub-chunk
  buffer.write("fmt ", 12); // Subchunk1ID
  buffer.writeUInt32LE(16, 16); // Subchunk1Size (16 for PCM)
  buffer.writeUInt16LE(1, 20); // AudioFormat (1 = PCM)
  buffer.writeUInt16LE(numChannels, 22); // NumChannels
  buffer.writeUInt32LE(sampleRate, 24); // SampleRate
  buffer.writeUInt32LE(byteRate, 28); // ByteRate
  buffer.writeUInt16LE(blockAlign, 32); // BlockAlign
  buffer.writeUInt16LE(bitsPerSample, 34); // BitsPerSample

  // data sub-chunk
  buffer.write("data", 36); // Subchunk2ID
  buffer.writeUInt32LE(dataSize, 40); // Subchunk2Size

  // Copy PCM samples
  Buffer.from(pcmBytes.buffer, pcmBytes.byteOffset, pcmBytes.byteLength).copy(buffer, 44);

  return buffer;
}

export function calculateWavDurationMs(pcmByteLength: number, sampleRate: number): number {
  const bytesPerSample = 2; // 16-bit mono
  const samples = pcmByteLength / bytesPerSample;
  return Math.round((samples / sampleRate) * 1000);
}

export const WavEncoder = {
  encodePcm16: encodePcm16ToWav,
  calculateDurationMs: calculateWavDurationMs,
};

File: android/app/src/main/java/com/feynmanlive/app/audio/WavEncoder.kt
package com.feynmanlive.app.audio

import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

object WavEncoder {

    fun encodePcm16(pcmBytes: ByteArray, sampleRate: Int = 16000, channels: Int = 1): ByteArray {
        val totalAudioLen = pcmBytes.size
        val totalDataLen = totalAudioLen + 36
        val byteRate = sampleRate * channels * 2

        val header = ByteArray(44)
        val buffer = ByteBuffer.wrap(header).order(ByteOrder.LITTLE_ENDIAN)

        // RIFF chunk descriptor
        buffer.put('R'.code.toByte())
        buffer.put('I'.code.toByte())
        buffer.put('F'.code.toByte())
        buffer.put('F'.code.toByte())
        buffer.putInt(totalDataLen)
        buffer.put('W'.code.toByte())
        buffer.put('A'.code.toByte())
        buffer.put('V'.code.toByte())
        buffer.put('E'.code.toByte())

        // fmt sub-chunk
        buffer.put('f'.code.toByte())
        buffer.put('m'.code.toByte())
        buffer.put('t'.code.toByte())
        buffer.put(' '.code.toByte())
        buffer.putInt(16) // SubChunk1Size for PCM
        buffer.putShort(1.toShort()) // AudioFormat (1 = PCM)
        buffer.putShort(channels.toShort())
        buffer.putInt(sampleRate)
        buffer.putInt(byteRate)
        buffer.putShort((channels * 2).toShort()) // BlockAlign
        buffer.putShort(16.toShort()) // BitsPerSample

        // data sub-chunk
        buffer.put('d'.code.toByte())
        buffer.put('a'.code.toByte())
        buffer.put('t'.code.toByte())
        buffer.put('a'.code.toByte())
        buffer.putInt(totalAudioLen)

        val out = ByteArrayOutputStream(44 + totalAudioLen)
        out.write(header)
        out.write(pcmBytes)
        return out.toByteArray()
    }

    fun savePcmAsWav(file: File, pcmBytes: ByteArray, sampleRate: Int = 16000, channels: Int = 1) {
        file.parentFile?.mkdirs()
        val wav = encodePcm16(pcmBytes, sampleRate, channels)
        FileOutputStream(file).use { it.write(wav) }
    }

    fun calculateDurationMs(pcmByteLength: Long, sampleRate: Int = 16000, channels: Int = 1): Long {
        val bytesPerSample = 2
        val bytesPerSec = sampleRate * channels * bytesPerSample
        return if (bytesPerSec > 0) (pcmByteLength * 1000L) / bytesPerSec else 0L
    }
}
